### DISCUSSION TITLE
Save workspaces locally

### DISCUSSION THREAD

**Discussion by McNull** [x1unix/go-playground#398] [Ideas]
I've spent some time today to save workspaces locally in the browser by workspace name. I think this makes it abit easier to quickly experiment with code and save it for later reference (like stackblitz or codesandbox).

![demo](https://github.com/user-attachments/assets/e60713f1-2cd0-4093-9f9e-9090ad6dcb7e)

The saved workspaces are stored in the browser by using indexDB and are not shared with other users. The code can still be shared by using the share button.

Let me know if you want a pull request. 

Thanks for this awesome tool!

**Comment by x1unix** [comment]
@McNull thanks for the idea and any PRs will be appreciated.

I like the idea, also using IndexedDB is a very good approach.

The only concern is Load button placement. Buttons space in the header is already a bit overwhelmed + layout also should fit on mobile devices.

Wdyt about putting workspace switcher to the right side as dropdown (similar to Go environment selector)?

Also right now, playground has autosave functionality that saves your changes unless snippet or example is open,
How do you handle autosave?

**Reply by McNull** [reply]
The button layout is indeed a bit overwhelmed. I think the workspace switcher as a dropdown is a good idea, will try to implement that.

I did change the autosave to always autosave, even when a snippet or example is loaded and changed. I actually didn't quite understand why it didn't autosave when a snippet was loaded, so I changed that. Maybe there a good reason but I thought the autosave was ment to autoreload the code where the user left it when they come back to the page. If someone makes changes to a shared snippet or example code they probably want that autosave that as well?
The autosave still makes use of the localstorage.

I will make a PR so you can easier see what I've done. I'm quite busy the next couple of days so it might take a while before I can finish it.
